### PR TITLE
BUG: pd.factorize should  not upconvert unique values unnecessarily

### PR DESCRIPTION
**Pull request by topper-123**
The uniques returned from `pd.factorize` should keep their original dtype, where possible.

```python
>>> arr =np.array([1,2], dtype="int8")
>>> pd.factorize(arr)
(array([0, 1], dtype=int64), array([1, 2], dtype=int64))  # master
(array([0, 1], dtype=int64), array([1, 2], dtype=int8))  # this PR
```

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -143,11 +143,14 @@ def _ensure_data(values: ArrayLike) -> tuple[np.ndarray, DtypeObj]:
             # until our algos support uint8 directly (see TODO)
             return np.asarray(values).astype("uint64"), np.dtype("bool")
         elif is_signed_integer_dtype(values):
-            return ensure_int64(values), np.dtype("int64")
+            dtype = getattr(values, "dtype", np.dtype("int64"))
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:146]: we should always have dtype at this point

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -143,11 +143,14 @@ def _ensure_data(values: ArrayLike) -> tuple[np.ndarray, DtypeObj]:
...
+            return ensure_int64(values), dtype
         elif is_unsigned_integer_dtype(values):
-            return ensure_uint64(values), np.dtype("uint64")
+            dtype = getattr(values, "dtype", np.dtype("uint64"))
+            return ensure_uint64(values), dtype
         elif is_float_dtype(values):
-            return ensure_float64(values), np.dtype("float64")
+            dtype = getattr(values, "dtype", np.dtype("float64"))
+            return ensure_float64(values), dtype
```
Comment (jbrockmendel) [inline pandas/core/algorithms.py:153]: i think this code was written before we had non-64 hashtables.  i expect a lot of this casting can now be avoided

File: pandas/core/algorithms.py (hunk 1 above)
Comment (topper-123) [reply pandas/core/algorithms.py]: Yeah, fixed.

File: pandas/core/algorithms.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/algorithms.py]: Probably yeah, and if we can avoid them, it'll speed some things up nicely.

But that's for a different PR?  I got a feeling there's a lot of detail hidden in that change.

Comment (jbrockmendel) [reply pandas/core/algorithms.py]: fair enough

**Comment by jreback** [comment]
thanks @topper-123

**Pull request merged**